fft.cpp

```diff
@@ -52,27 +54,31 @@ void FFT(cpx *in, cpx *out, int step, int size, int dir)
     out[0] = in[0];
     return;
   }
-  FFT(in, out, step * 2, size / 2, dir);
-  FFT(in + step, out + size / 2, step * 2, size / 2, dir);
-  for(int i = 0 ; i < size / 2 ; i++)
+  int half = size >> 1;
+  FFT(in, out, step * 2, half, dir);
+  FFT(in + step, out + half, step << 1, half, dir);
+  for(int i = 0 ; i < half ; i++)
   {
     cpx even = out[i];
-    cpx odd = out[i + size / 2];
+    cpx odd = out[i + half];
     out[i] = even + EXP(dir * two_pi * i / size) * odd;
-    out[i + size / 2] = even + EXP(dir * two_pi * (i + size / 2) / size) * odd;
+    out[i + half] = even + EXP(dir * two_pi * (i + half) / size) * odd;
   }
 }
 
-// Usage:
-// f[0...N-1] and g[0..N-1] are numbers
-// Want to compute the convolution h, defined by
-// h[n] = sum of f[k]g[n-k] (k = 0, ..., N-1).
-// Here, the index is cyclic; f[-1] = f[N-1], f[-2] = f[N-2], etc.
-// Let F[0...N-1] be FFT(f), and similarly, define G and H.
-// The convolution theorem says H[n] = F[n]G[n] (element-wise product).
-// To compute h[] in O(N log N) time, do the following:
-//   1. Compute F and G (pass dir = 1 as the argument).
-//   2. Get H by element-wise multiplying F and G.
-//   3. Get h by taking the inverse FFT (use dir = -1 as the argument)
-//      and *dividing by N*. DO NOT FORGET THIS SCALING FACTOR.
-// To compute an *acyclic* convolution, pad f and g to the right with zeroes.
+// f:      input array f
+// g:      input array g
+// h:      output array h
+// size:   length of f/g/h {MUST BE A POWER OF 2}
+// RESULT: h[i] = sum of f[k]*g[i-k] for k=0..(size-1)
+// Note: f and g are cyclic (e.g. f[-1] = f[size-1]), so they may be padded
+// the right with zeroes to instead compute an acyclic convolution.
+void FFTConvolution(cpx *f, cpx *g, cpx *h, int size)
+{
+  cpx *F = new cpx[size], *G = new cpx[size], *H = new cpx[size];

```

Comment:
can we use `std::vector<cpx>` to avoid the memory leak?

Reply:
Rewriting it all to use vectors doesn't seem to work well, but I guess we can manually free the arrays instead.

Reply:
does this not work?

std::vector<cpx> F(size), G(size), H(size);
FFT(f, F.data(), 1, size, 1);
FFT(g, G.data(), 1, size, 1);
// no change to line 81
FFT(H.data(), h, size, -1);
// no change to line 83

Reply:
Looks like that's unsupported by older compilers, so I'd rather not do it that way.